core/src/main/java/io/grpc/internal/AutoConfiguredLoadBalancerFactory.java

```diff
@@ -214,59 +216,82 @@ PolicySelection decideLoadBalancerProvider(
-            logger.warning(errorMsg);
-          }
-          return new PolicySelection(
-              getProviderOrThrow(
-                  "round_robin", "received balancer addresses but grpclb runtime is missing"),
-              backendAddrs, null);
-        } else {
-          return new PolicySelection(grpclbProvider, servers, null);
-        }
-      }
-      roundRobinDueToGrpclbDepMissing = false;
-
       List<LbConfig> lbConfigs = null;
       if (config != null) {
         List<Map<String, Object>> rawLbConfigs =
             ServiceConfigUtil.getLoadBalancingConfigsFromServiceConfig(config);
         lbConfigs = ServiceConfigUtil.unwrapLoadBalancingConfigList(rawLbConfigs);
       }
       if (lbConfigs != null && !lbConfigs.isEmpty()) {
+        // If there isn't a preselecbted, try to get the provider as configured
         LinkedHashSet<String> policiesTried = new LinkedHashSet<>();
         for (LbConfig lbConfig : lbConfigs) {
           String policy = lbConfig.getPolicyName();
-          LoadBalancerProvider provider = registry.getProvider(policy);
-          if (provider != null) {
-            if (!policiesTried.isEmpty()) {
-              helper.getChannelLogger().log(
-                  ChannelLogLevel.DEBUG,
-                  "{0} specified by Service Config are not available", policiesTried);
+          if (policy.equals(GRPCLB_POLICY_NAME)) {
+            return selectGrpclb(servers, backendAddrs, lbConfig.getRawConfigValue());
+          } else {
+            LoadBalancerProvider provider = registry.getProvider(policy);
+            if (provider == null) {
+              policiesTried.add(policy);
+            } else {
+              if (!policiesTried.isEmpty()) {
+                // Before returning, log all previously tried policies
+                helper.getChannelLogger().log(
+                    ChannelLogLevel.DEBUG,
+                    "{0} specified by Service Config are not available", policiesTried);
+              }
+              return new PolicySelection(provider, backendAddrs, lbConfig.getRawConfigValue());
             }
-            return new PolicySelection(provider, servers, lbConfig.getRawConfigValue());
           }
-          policiesTried.add(policy);
         }
-        throw new PolicyException(
+        if (!haveBalancerAddress) {
+          throw new PolicyException(
             "None of " + policiesTried + " specified by Service Config are available.");
+        }
+      }
+
+      if (haveBalancerAddress) {
+        // This is a special case where the existence of balancer address in the resolved address
+        // selects "grpclb" policy if the service config couldn't select a policy
+        return selectGrpclb(servers, backendAddrs, null);
       }
+      // No config nor balancer address. Use default.
       return new PolicySelection(
           getProviderOrThrow(defaultPolicy, "using default policy"), servers, null);
     }
+
+    private PolicySelection selectGrpclb(
```

Comment:
This looks like it should only be used in the `if (haveBalancerAddress) {` case.